src/pages/groupComparison/vennDiagram/VennUtils.ts

```diff
@@ -0,0 +1,386 @@
+                size -= rectangleArea(rectangleIntersection(rectA, rectC));
+                size += rectangleArea(rectangleIntersection(..._.values(setRectangles)));
+            }
+        }
+    }
+
+    return size;
+}
+
+function isPointInsideRectangle(x:number, y:number, rectangle:Rectangle) {
+    return ((x > rectangle.x) &&
+        (x < rectangle.x + rectangle.xLength) &&
+        (y > rectangle.y) &&
+        (y < rectangle.y + rectangle.yLength));
+}
+
+function isPointInsideRegion(x:number, y:number, sets:string[], excludedSets:string[], setRectangles:SetRectangles) {
+    let inside = true;
+    for (const set of sets) {
+        inside = inside && isPointInsideRectangle(x, y, setRectangles[set]);
+    }
+    for (const set of excludedSets) {
+        inside = inside && !isPointInsideRectangle(x, y, setRectangles[set]);
+    }
+    return inside;
+}
+
+export function getApproximateRegionArea(
+    sets:string[],
+    setRectangles:SetRectangles
+) {
+    // for testing
+    const xMin = Math.min(...sets.map(setId=>setRectangles[setId].x));
+    const xMax = Math.max(...sets.map(setId=>setRectangles[setId].x+setRectangles[setId].xLength));
+    const yMin = Math.min(...sets.map(setId=>setRectangles[setId].y));
+    const yMax = Math.max(...sets.map(setId=>setRectangles[setId].y+setRectangles[setId].yLength));
+
+    const numSamples = 2000000;
+    let numInside = 0;
+    const excludedSets = _.difference(Object.keys(setRectangles), sets);
+    for (let i=0; i<numSamples; i++) {
+        const sampleX = Math.random()*(xMax-xMin) + xMin;
+        const sampleY = Math.random()*(yMax-yMin) + yMin;
+        numInside += +isPointInsideRegion(sampleX, sampleY, sets, excludedSets, setRectangles);
+    }
+
+    return (numInside / numSamples) * (xMax - xMin) * (yMax - yMin);
+}
+
+export function lossFunction(
+    setRectangles:SetRectangles,
+    areas:Area[]
+) {
+    // NOTE: it seems like zeroAreaPenalty and rectangleRatioPenalty are not necessary, so I've commented them out,
+    //      but am leaving them commented in case we run into edge cases where those are an issue.
+    let areaError = 0;
+    let intersectionDistancePenalty = 0;
+
+    /* UNCOMMENT IF ZERO AREAS SHOW UP
+        let areasThatShouldBeZero = 0;
+     */
+    for (const area of areas) {
+        // Make regions proportional to their size
+        const regionArea = getRegionArea(area.sets, setRectangles);
+        const error = (regionArea - area.preciseSize);
+        areaError += error*error;
+
+        if (area.sets.length === 1) {
+            // Make each rectangle proportional to its size
+            const rectSize = rectangleArea(setRectangles[area.sets[0]]);
+            const rectSizeError = (rectSize - area.size);
+            areaError += rectSizeError*rectSizeError;
+        } else if (area.sets.length === 2 && regionArea === 0 && area.preciseSize > 0) {
+            // try to bring each pair of rectangles together if they should intersect
```

Comment:
clarify `if they should intersect and currently dont. Otherwise the loss which only detects errors in intersection areas won't know which direction to move them to get better.`